platform/solr/solr-factory/src/main/java/org/codice/solr/factory/impl/SolrClientAdapter.java

```diff
@@ -0,0 +1,998 @@
+        // if we get here than the ping was successful so make sure we move to connected
+        setConnected(true);
+        // fall-through to return the current one which should have been changed to the actual one
+      } catch (UnavailableSolrException e) {
+        // fall-through to return the current one which should still be an unavailable one
+      }
+    }
+    return this.apiClient;
+  }
+
+  @Override
+  protected <T> T handle(Code<T> code) throws SolrServerException, IOException {
+    try {
+      return code.invoke(getProxiedClient());
+    } catch (UnavailableSolrException e) {
+      throw e;
+    } catch (SolrException e) {
+      LOGGER.debug(
+          "Solr({}): API failure with code [{}] and metadata [{}]; {}",
+          core,
+          e.code(),
+          e.getMetadata(),
+          e, // this will get the info logged on the first line
+          e); // this one will get the stack trace logged after
+      checkIfReachableAndChangeStateAccordingly(e, "from the API after an error was detected");
+      throw e;
+    } catch (SolrServerException | IOException e) {
+      LOGGER.debug("Solr({}): API failure; {}", core, e, e);
+      checkIfReachableAndChangeStateAccordingly(e, "from the API after an error was detected");
+      throw e;
+    }
+  }
+
+  @Override
+  public final SolrClient getClient() {
+    // returning this to make sure all calls from our SolrClient API will still be intercepted
+    // when the client is retrieved and passed to a Solr request object
+    // this should be temporary until we completely abstract out Solr with interfaces
+    return this;
+  }
+
+  @Override
+  public String getCore() {
+    return core;
+  }
+
+  @Override
+  @SuppressWarnings("squid:S2093" /* closing of real client handled by finalizeStateChange() */)
+  public void close() throws IOException {
+    try {
+      if (state == State.CLOSED) {
+        return;
+      }
+      final SolrClient previousClientToClose;
+      final Future<?> futureToCancel;
+
+      synchronized (lock) {
+        if (state == State.CLOSED) { // already closed so bail
+          return;
+        }
+        futureToCancel = future;
+        previousClientToClose = realClient;
+        LOGGER.debug("Solr({}): closing", core);
+        this.unavailableClient =
+            new UnavailableSolrClient(
+                new UnavailableSolrException("'" + core + "' client was closed"));
+        this.apiClient = unavailableClient;
+        this.pingClient = unavailableClient;
+        this.realClient = null;
+        this.state = State.CLOSED;
+        this.future = null;
+        lock.notifyAll(); // wakeup those waiting for isAvailable(timeout)
+      }
+      finalizeStateChange(true, futureToCancel, previousClientToClose, false);
+    } finally {
+      listeners.clear();
+      initializers.clear();
+    }
+  }
+
+  @Override
+  public boolean isAvailable() {
+    final boolean available = (state == State.CONNECTED);
```

Comment:
That would require me to synchronize which I was trying to avoid here. Since by design, this class is dealing with background retries when it is not available, this method doesn't have to do anything, As such, checking only for `CONNECTED` would automatically yield `false` and return right away without doing anything else; very close to a short-circuit.

Reply:
Good reason. Would the tests fail if someone were to change the way this is done? If not, we may want to add a comment to make sure this doesn't get changed by mistake.

Reply:
added the above explanation to the code